libyul/backends/evm/EVMDialect.cpp

```diff
@@ -387,27 +397,43 @@ EVMDialect::EVMDialect(langutil::EVMVersion _evmVersion, std::optional<uint8_t>
 	m_functions(createBuiltins(_evmVersion, _eofVersion, _objectAccess)),
 	m_reserved(createReservedIdentifiers(_evmVersion))
 {
+	m_handles.add = EVMDialect::builtin("add");
+	m_handles.exp = EVMDialect::builtin("exp");
+	m_handles.mul = EVMDialect::builtin("mul");
+	m_handles.not_ = EVMDialect::builtin("not");
+	m_handles.shl = EVMDialect::builtin("shl");
+	m_handles.sub = EVMDialect::builtin("sub");
+
+	m_discardFunction = EVMDialect::builtin("pop");
+	m_equalityFunction = EVMDialect::builtin("eq");
+	m_booleanNegationFunction = EVMDialect::builtin("iszero");
+	m_memoryStoreFunction = EVMDialect::builtin("mstore");
+	m_memoryLoadFunction = EVMDialect::builtin("mload");
+	m_storageStoreFunction = EVMDialect::builtin("sstore");
+	m_storageLoadFunction = EVMDialect::builtin("sload");
+	m_hashFunction = EVMDialect::builtin("keccak256");
 }
 
-BuiltinFunctionForEVM const* EVMDialect::builtin(YulName _name) const
+std::optional<BuiltinHandle> EVMDialect::builtin(std::string_view _name) const
 {
 	if (m_objectAccess)
 	{
 		std::smatch match;
-		if (regex_match(_name.str(), match, verbatimPattern()))
+		std::string name(_name);
+		if (regex_match(name, match, verbatimPattern()))
 			return verbatimFunction(stoul(match[1]), stoul(match[2]));
 	}
-	auto it = m_functions.find(_name);
-	if (it != m_functions.end())
-		return &it->second;
-	else
-		return nullptr;
+	auto it = std::find_if(m_functions.begin(), m_functions.end(), [&_name](auto const& builtin) { return builtin && builtin.value().name == _name; });
+	if (it != m_functions.end() && *it && it->value().name == _name)
+		// ids are offset by the maximum number of verbatim functions
+		return BuiltinHandle{static_cast<size_t>(std::distance(m_functions.begin(), it)) + verbatimIdOffset};
+	return std::nullopt;
```

Comment:
This changes `builtin()` from O(1) to O(n) and now it's also comparing `string`s rather than `YulString`s. Maybe we should also add a helper `map` to make lookups faster? Especially given that `Dialect`s are created once and then used a lot.

`builtin()` does not seem to be used that much compared to `builtinFunction()`, but there are a few uses in the parser and analysis that will run basically on every identifier in the source. Since we do a lot of parsing and reparsing internally, it could add up. I saw parsing taking around 6% of total runtime for Uniswap last time I tried profiling it (https://github.com/ethereum/solidity/pull/15451#issuecomment-2372539583), which makes me a bit worried about this. Perhaps it's worth trying out and checking if it affects compilation time.

Reply:
Hmmmm yeah, that is a good point. I'll add a helper map, that should help with parser and analysis performance.

Reply:
> (although it's O(logn) to O(n), I'd think).

True. I'm usually thinking about associative arrays as having amortized O(1) lookups, but I forget that not all of them are implemented as hash tables. In C++ that would be `unordered_map` instead. `unordered_map` actually looks perfect for this case. We're reading much more than writing, so faster lookups are preferable.

Reply:
**__**

Reply:
And that's just the complexity, another is the efficiency of things in relation to the data we're accessing. Could be that a linear vector with just searching through it is in practice more efficient (albeit worse complexity-wise) than an unordered map lookup. Remains to be seen!

Reply:
Could be. I'm bringing this up because it looks like it might be a bottleneck, but ultimately we should of course measure this and verify.